packages/protocol/src/utils/StreamPartitionID.ts

```diff
@@ -0,0 +1,154 @@
+/**
+ * SPID - Stream Partition ID
+ */
+export default class SPID {
+    /** stream id */
+    public readonly id: string
+    /** stream partition */
+    public readonly partition: number
+    /** toString/fromString separator */
+    protected static readonly SEPARATOR = '|'
+
+    /** string key representing SPID */
+    public readonly key: string
+
+    protected readonly cachedSplit: [string, number] // cached split
+
+    /**
+     * @param id - stream id
+     * @param partition - stream partition
+     */
+    constructor(id: string, partition = 0) {
+        this.id = typeof id === 'string' ? id.toLowerCase() : id
+        this.partition = partition
+        this.validate()
+
+        // static values
+        const separator = (this.constructor as typeof SPID).SEPARATOR
+        this.key =`${this.id}${separator}${this.partition}`
+        this.cachedSplit = [this.id, this.partition]
+        // prevent all mutation
+        Object.freeze(this)
+        Object.freeze(this.cachedSplit)
+    }
+
+    /**
+     * Throws if data is invalid.
+     */
+    private validate() {
+        if (typeof this.id !== 'string' || this.id.length === 0) {
+            throw new Error('SPID validation failed: id must be non-empty string')
+        }
+
+        if (typeof this.partition !== 'number' || !Number.isSafeInteger(this.partition) || this.partition < 0) {
+            throw new Error('SPID validation failed: paritition must be an integer >= 0 ')
+        }
+    }
+
+    /**
+     * True iff other value is equivalent.
+     */
+    equals(other: SPIDish): boolean {
+        // check if same instance
+        if (other === this) { return true }
+        let otherSpid: SPID
+        try {
+            otherSpid = (this.constructor as typeof SPID).from(other)
+        } catch (_err) {
+            // not equal if not valid
+            return false
+        }
+
+        // check if matches both id + partition
+        return (
+            this.id === otherSpid.id
+            && this.partition === otherSpid.partition
+        )
+    }
+
+    /**
+     * Convert to SPID if possible.
+     */
+    static from(spidish: SPIDish): SPID {
+        // return spid if already spid
+        if (spidish instanceof SPID) {
+            return spidish
+        }
+
+        // convert from string
+        if (typeof spidish === 'string') {
+            return SPID.fromString(spidish)
+        } else if (spidish && typeof spidish === 'object') {
+            // @ts-expect-error object should have one of these, validated anyway
+            const streamId = spidish.streamId || spidish.id
+            // @ts-expect-error object should have one of these, validated anyway
+            const streamPartition = spidish.streamPartition || spidish.partition
+            return SPID.from(`${streamId}${this.SEPARATOR}${streamPartition}`)
```

Comment:
More direct way of constructing instance `return new SPID(streamId, streamPartition)` to avoid redundant string parsing.

Reply:
I think the intent here was to give partition a chance to run through parseFloat but yeah perhaps doesn't matter.